bundles/sirix-core/src/main/java/org/sirix/service/xml/shredder/WikipediaImport.java

```diff
@@ -0,0 +1,588 @@
+import javax.xml.stream.XMLInputFactory;
+import javax.xml.stream.XMLStreamConstants;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.events.Attribute;
+import javax.xml.stream.events.Namespace;
+import javax.xml.stream.events.StartElement;
+import javax.xml.stream.events.XMLEvent;
+import org.sirix.access.DatabaseConfiguration;
+import org.sirix.access.Databases;
+import org.sirix.access.ResourceConfiguration;
+import org.sirix.api.Axis;
+import org.sirix.api.Database;
+import org.sirix.api.NodeTrx;
+import org.sirix.api.xml.XmlNodeReadOnlyTrx;
+import org.sirix.api.xml.XmlNodeTrx;
+import org.sirix.api.xml.XmlResourceManager;
+import org.sirix.diff.algorithm.fmse.DefaultNodeComparisonFactory;
+import org.sirix.diff.algorithm.fmse.FMSE;
+import org.sirix.exception.SirixException;
+import org.sirix.node.NodeKind;
+import org.sirix.service.xml.xpath.XPathAxis;
+import org.sirix.utils.LogWrapper;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <h1>WikipediaImport</h1>
+ *
+ * <p>
+ * Import sorted Wikipedia revisions. Precondition is a file, which is produced from a Hadoop job.
+ * </p>
+ *
+ * @author Johannes Lichtenberger, University of Konstanz
+ *
+ */
+public final class WikipediaImport implements Import<StartElement> {
+
+  /** {@link LogWrapper} reference. */
+  private static final LogWrapper LOGWRAPPER = new LogWrapper(LoggerFactory.getLogger(WikipediaImport.class));
+
+  /** StAX parser {@link XMLEventReader}. */
+  private transient XMLEventReader mReader;
+
+  /** Resource manager instance. */
+  private final XmlResourceManager mResourceManager;
+
+  /** sirix {@link NodeTrx}. */
+  private transient XmlNodeTrx mWtx;
+
+  /** {@link XMLEvent}s which specify the page metadata. */
+  private transient Deque<XMLEvent> mPageEvents;
+
+  /** Determines if page has been found in sirix storage. */
+  private transient boolean mFound;
+
+  /** String value of text-element. */
+  private transient String mIdText;
+
+  /** Determines if StAX parser is currently parsing revision metadata. */
+  private transient boolean mIsRev;
+
+  /** Timestamp of each revision as a simple String. */
+  private transient String mTimestamp;
+
+  /** Sirix {@link Database}. */
+  private final Database<XmlResourceManager> mDatabase;
+
+  /** Revision-timespan by date. */
+  public enum DateBy {
+    /** By seconds. */
+    SECONDS,
+
+    /** By minutes. */
+    MINUTES,
+
+    /** By hours. */
+    HOURS,
+
+    /** By days. */
+    DAYS,
+
+    /** By months. */
+    MONTHS
+  }
+
+  /**
+   * Constructor.
+   *
+   * @param xmlFile The XML file to import.
+   * @param sirixDatabase The sirix destination storage directory.
+   *
+   */
+  public WikipediaImport(final Path xmlFile, final Path sirixDatabase) throws SirixException {
+    mPageEvents = new ArrayDeque<>();
+    final XMLInputFactory xmlif = XMLInputFactory.newInstance();
+    try {
+      mReader = xmlif.createXMLEventReader(new FileInputStream(checkNotNull(xmlFile.toFile())));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor call: *FileInputStream*. Currently, there are execution paths that do not contain closure statements, e.g., when exception is thrown by XMLInputFactory.createXMLEventReader. Either a) close the object returned by *FileInputStream()* in a try-finally block or b) close the resource by declaring the object returned by *FileInputStream()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).